File: js/test/layout.test.ts
import { describe, expect, it } from "vitest";
import {
  Engine,
  LayoutRenderer,
  RendererLayout,
  renderToPdf,
  type PdfConverter,
} from "../src/index.js";

describe("layout renderer", () => {
  it("renders blocks to HTML", () => {
    const e = new Engine();
    e.registerRenderer(new LayoutRenderer());
    const src = JSON.stringify({
      style: { accent: "#123456" },
      blocks: [
        { type: "heading", text: "{{ title }}" },
        { type: "paragraph", text: "Hi {{ name }}" },
        { type: "button", text: "Open", url: "{{ url }}" },
      ],
    });
    e.register({ name: "promo", renderer: RendererLayout, parts: { html: src } });
    const out = e.renderPart("promo", "html", { title: "Sale", name: "Ada", url: "https://x.test" });
    expect(out).toContain(`<h1 class="guten-heading" style="color:#123456">Sale</h1>`);
    expect(out).toContain(`<p class="guten-paragraph">Hi Ada</p>`);
    expect(out).toContain(`<a class="guten-button" href="https://x.test" style="background:#123456">Open</a>`);
  });
});

describe("pdf seam", () => {
  it("renders html and delegates to the converter", async () => {
    const e = new Engine();
    e.register({ name: "doc", parts: { html: "<h1>{{ t }}</h1>" } });
    let got = "";
    const conv: PdfConverter = {
      toPdf: (html) => {
        got = html;
        return new TextEncoder().encode("%PDF-1.4\n" + html);
      },
    };
    const pdf = await renderToPdf(e, "doc", { t: "Hi" }, conv);
    expect(new TextDecoder().decode(pdf).startsWith("%PDF")).toBe(true);
    expect(got).toContain("<h1>Hi</h1>");
  });

  it("errors without a converter", async () => {
    const e = new Engine();
    e.register({ name: "doc", parts: { html: "x" } });
    await expect(
      renderToPdf(e, "doc", {}, undefined as unknown as PdfConverter),
    ).rejects.toThrow();
  });
});

File: go/layout_test.go
package guten

import (
	"context"
	"errors"
	"strings"
	"testing"
)

func TestLayoutRenderer(t *testing.T) {
	e := New()
	e.RegisterRenderer(NewLayoutRenderer())
	src := `{"style":{"accent":"#123456"},"blocks":[{"type":"heading","text":"{{ title }}"},{"type":"paragraph","text":"Hi {{ name }}"},{"type":"button","text":"Open","url":"{{ url }}"}]}`
	if err := e.Register(Template{Name: "promo", Renderer: RendererLayout, Parts: map[string]string{PartHTML: src}}); err != nil {
		t.Fatalf("register: %v", err)
	}
	out, err := e.RenderPart("promo", PartHTML, map[string]any{"title": "Sale", "name": "Ada", "url": "https://x.test"})
	if err != nil {
		t.Fatalf("render: %v", err)
	}
	for _, want := range []string{
		`<h1 class="guten-heading" style="color:#123456">Sale</h1>`,
		`<p class="guten-paragraph">Hi Ada</p>`,
		`<a class="guten-button" href="https://x.test" style="background:#123456">Open</a>`,
	} {
		if !strings.Contains(out, want) {
			t.Fatalf("missing %q in:\n%s", want, out)
		}
	}
}

// fakePDF stands in for an injected PDF converter (the seam consumers wire).
type fakePDF struct {
	err error
	got []byte
}

func (f *fakePDF) ToPDF(_ context.Context, html []byte) ([]byte, error) {
	if f.err != nil {
		return nil, f.err
	}
	f.got = html
	return append([]byte("%PDF-1.4\n"), html...), nil
}

func TestRenderToPDF(t *testing.T) {
	e, _ := NewWithBuiltins()
	conv := &fakePDF{}
	pdf, err := e.RenderToPDF(context.Background(), "basic_notification", map[string]any{"title": "Hi", "body": "x"}, conv)
	if err != nil {
		t.Fatalf("RenderToPDF: %v", err)
	}
	if !strings.HasPrefix(string(pdf), "%PDF") {
		t.Fatalf("not pdf: %q", string(pdf))
	}
	if !strings.Contains(string(conv.got), "<h1") {
		t.Fatal("converter did not receive rendered html")
	}
	if _, err := e.RenderToPDF(context.Background(), "basic_notification", nil, nil); err == nil {
		t.Fatal("expected error for nil converter")
	}
	if _, err := e.RenderToPDF(context.Background(), "basic_notification", nil, &fakePDF{err: errors.New("boom")}); err == nil {
		t.Fatal("expected converter error")
	}
}
